### DISCUSSION TITLE
Which Faces Versions to Support for 5.0?

### DISCUSSION THREAD

**Discussion by larsgrefer** [joinfaces/joinfaces#1516] [Polls]
Should we support both Faces 3 (EE9) and Faces 4 (EE10) or should we use Faces 4 as minimum version and don't support Faces 3?

**Comment by persapiens** [comment]
Let me talk a little about migration's perspective.

When I think about migrating apps from Java EE 8 (joinfaces 4.7.x) to Jakarta EE 10 (joinfaces 5.0.x), I'll have to make at least two main changes to my apps:

- Change namespaces from **javax** to **jakarta**: Both Jakarta EE 9 and Jakarta EE 10 require this change.
- Use full CDI implementation: Jakarta EE 9 does not require this change yet. Jakarta EE 10 does.

Option "Faces 3+4 / Jakarta EE 9+10" enables incremental migration.

**Comment by persapiens** [comment]
@larsgrefer ,

[Spring Boot 3 has just changed the default version to Jakarta EE 10.](https://spring.io/blog/2022/11/10/spring-boot-3-0-0-rc2-available-now)

More, Jakarta EE 9 support is deprecating fast. For example, [The Apache Tomcat team announces support for Apache Tomcat 10.0.x ended on 31 October 2022.](https://tomcat.apache.org/tomcat-10.0-eol.html) Tomcat 10 supports some Jakarta EE 9 specifications.

In my previous message, I preferred "incremental migration." But these two announcements changed my mind. So, I prefer the "Faces 4 / Jakarta EE 10 only" option now.

**Reply by larsgrefer** [reply]
Supporting Only Faces 4 would be easier. But Jetty support seems to be broken since Spring Boot 3.0.0-RC2.

Let's wait for https://github.com/spring-projects/spring-boot/issues/33044

**Comment by larsgrefer** [comment]
@persapiens 

For JoinFaces 5.x, we kept support for Faces 3 and Faces 4

With JoinFaces 6 (Spring Boot 4), we could now drop Faces 3 support and concentrate on Faces 4.0/4.1

**Comment by persapiens** [comment]
Hi @larsgrefer 

[Spring Boot 4 will support Servlet 6.1+ compatible container.](https://docs.spring.io/spring-boot/4.0/system-requirements.html)

[Faces 3.0 supports Servlet 5.0 / Jakarta EE 9](https://central.sonatype.com/artifact/jakarta.faces/jakarta.faces-api/3.0.0/dependencies)
[Faces 4.0 supports Servlet 6.0 / Jakarta EE 10](https://central.sonatype.com/artifact/jakarta.faces/jakarta.faces-api/4.0.1/dependencies)
[Faces 4.1 supports Servlet 6.1 / Jakarta EE 11](https://central.sonatype.com/artifact/jakarta.faces/jakarta.faces-api/dependencies)

I agree that we should drop support for Faces 3.0 and focus on Faces 4.0/4.1 instead.  
Servlet 6.1+ containers run Faces 4.1, and they will also run Faces 4.0 since it requires Servlet 6.0 — a minor version difference.
